File: frontend/src/types/settlement.ts
import type {
  MoneyValue
} from "./expense";


export interface Settlement {
  id: string;
  group_id: string;
  from_user_id: string;
  from_user_name: string;
  to_user_id: string;
  to_user_name: string;
  amount: MoneyValue;
  currency: string;
  note: string | null;
  settlement_date: string;
  created_by_id: string;
  created_at: string;
}


export interface CreateSettlementPayload {
  from_user_id: string;
  to_user_id: string;
  amount: number;
  note: string;
  settlement_date: string;
}

File: backend/app/schemas/settlement.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import (
    BaseModel,
    Field,
    model_validator
)


class SettlementCreate(BaseModel):
    from_user_id: UUID
    to_user_id: UUID

    amount: Decimal = Field(
        gt=0,
        max_digits=12,
        decimal_places=2
    )

    note: str | None = Field(
        default=None,
        max_length=300
    )

    settlement_date: date = Field(
        default_factory=date.today
    )

    @model_validator(mode="after")
    def validate_different_users(
        self
    ):
        if (
            self.from_user_id ==
            self.to_user_id
        ):
            raise ValueError(
                "Sender and receiver must be different"
            )

        return self


class SettlementResponse(BaseModel):
    id: UUID
    group_id: UUID
    from_user_id: UUID
    from_user_name: str
    to_user_id: UUID
    to_user_name: str
    amount: Decimal
    currency: str
    note: str | None
    settlement_date: date
    created_by_id: UUID
    created_at: datetime